[Bug] trandom实现中一系列不支持ui32的接口

### Component

PTO Dialect / ODS (include/PTO/IR)

### Description

pto-isa中就是用的ui32，如果ptoas中改为i32，则会报错超出最大值
TrandomOp
<img width="865" height="610" alt="Image" src="https://github.com/user-attachments/assets/3a37434b-af83-4e1f-8d99-08374703e250" />

pto.load_scalar
<img width="1249" height="226" alt="Image" src="https://github.com/user-attachments/assets/bcbb0c72-98b4-45d5-a2b2-e35fe9c555db" />


### Reproduction (minimal)

```markdown
python3 test/tilelang_st/script/run_st.py -r sim -v a5 -t trandom
```

### Expected behavior

pass

### Actual behavior / error logs

```text
// === [PTOInsertSync] End === //


  loc("/home/wujiajun/llvm-workspace/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/trandom/trandom.pto":40:13):
  error: failed to legalize operation 'pto.load_scalar' that was explicitly
  marked illegal

  VPTO LLVM emission failed: VPTO type lowering failed

  VPTO LLVM emission failed: lowerVPTOTypes failed

  Error: Failed to emit VPTO text.


make[2]: *** [testcase/trandom/CMakeFiles/trandom_kernel.dir/build.make:85: testcase/trandom/trandom_kernel.ll] Error 1
make[2]: Leaving directory '/home/wujiajun/llvm-workspace/PTOAS/test/tilelang_st/npu/a5/src/st/build'
make[1]: *** [CMakeFiles/Makefile2:114: testcase/trandom/CMakeFiles/trandom_kernel.dir/all] Error 2
make[1]: Leaving directory '/home/wujiajun/llvm-workspace/PTOAS/test/tilelang_st/npu/a5/src/st/build'
make: *** [Makefile:91: all] Error 2
run failed: Command '['make', 'VERBOSE=1', '-j', '64']' returned non-zero exit status 2.
```

### Git commit

cb278a6ebc52f5cd1c441065abf0df7844a74662

### Host platform

None

### Target Ascend arch (if relevant)

None

### PTOAS build level (if relevant)

None

--- a/lib/PTO/Transforms/VPTOLLVMEmitter.cpp
+++ b/lib/PTO/Transforms/VPTOLLVMEmitter.cpp
@@ -6570,6 +6570,12 @@ public:
     if (!llvmPtrType)
       return rewriter.notifyMatchFailure(op, "expected LLVM pointer operand");
 
+    Type convertedValueType =
+        getTypeConverter()->convertType(op.getValue().getType());
+    if (!convertedValueType)
+      return rewriter.notifyMatchFailure(op,
+                                         "could not convert load_scalar result type");
+
     Value offset = adaptor.getOffset();
     if (offset.getType().isIndex())
       offset = rewriter.create<arith::IndexCastUIOp>(op.getLoc(),
@@ -6578,7 +6584,7 @@ public:
     Value elemPtr = adaptor.getPtr();
     if (!matchPattern(offset, m_Zero())) {
       elemPtr = rewriter.create<LLVM::GEPOp>(op.getLoc(), llvmPtrType,
-                                             op.getValue().getType(), adaptor.getPtr(),
+                                             convertedValueType, adaptor.getPtr(),
                                              ValueRange{offset});
     }
 
@@ -6596,8 +6602,8 @@ public:
     };
 
     rewriter.replaceOpWithNewOp<LLVM::LoadOp>(
-        op, op.getValue().getType(), elemPtr,
-        getNaturalAlignment(op.getValue().getType()));
+        op, convertedValueType, elemPtr,
+        getNaturalAlignment(convertedValueType));
     return success();
   }
 };

--- a/lib/PTO/Transforms/VPTOPtrNormalize.cpp
+++ b/lib/PTO/Transforms/VPTOPtrNormalize.cpp
@@ -1,3 +1,11 @@
+// Copyright (c) 2026 Huawei Technologies Co., Ltd.
+// This program is free software, you can redistribute it and/or modify it under the terms and conditions of
+// CANN Open Software License Agreement Version 2.0 (the "License").
+// Please refer to the License for details. You may not use this file except in compliance with the License.
+// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
+// See LICENSE in the root of the software repository for the full text of the License.
+
 #include "PTO/IR/PTO.h"
 #include "PTO/Transforms/Passes.h"
 
@@ -152,6 +160,64 @@ static Value materializeSubviewInputPtr(Value source, PatternRewriter &rewriter,
   auto ptrType = pto::PtrType::get(rewriter.getContext(),
                                    memrefType.getElementType(), memorySpace);
   return rewriter.create<pto::CastPtrOp>(loc, ptrType, source);
+}
+
+static Value materializeScalarAccessPtr(Value source, PatternRewriter &rewriter,
+                                        Location loc) {
+  if (!source)
+    return {};
+  if (isa<pto::PtrType>(source.getType()))
+    return source;
+
+  if (auto cast = source.getDefiningOp<UnrealizedConversionCastOp>()) {
+    if (cast->getNumOperands() != 1 || cast->getNumResults() != 1)
+      return {};
+    Value input = cast.getOperands().front();
+    if (isa<pto::PtrType>(input.getType()))
+      return input;
+    return materializeScalarAccessPtr(input, rewriter, loc);
+  }
+
+  if (auto cast = source.getDefiningOp<memref::CastOp>())
+    return materializeScalarAccessPtr(cast.getSource(), rewriter, loc);
+
+  if (auto subview = source.getDefiningOp<memref::SubViewOp>()) {
+    if (!needsSubviewPtrConversion(subview))
+      return {};
+
+    Value basePtr =
+        materializeScalarAccessPtr(subview.getSource(), rewriter, loc);
+    if (!basePtr)
+      return {};
+
+    Value offset;
+    if (failed(computeSubviewElementOffset(subview, rewriter, offset)))
+      return {};
+
+    auto ptrType = dyn_cast<pto::PtrType>(convertSubviewResultType(source.getType()));
+    if (!ptrType)
+      return {};
+    if (basePtr.getType() != ptrType)
+      basePtr = rewriter.create<pto::CastPtrOp>(loc, ptrType, basePtr);
+    return rewriter.create<pto::AddPtrOp>(loc, ptrType, basePtr, offset);
+  }
+
+  if (auto bind = source.getDefiningOp<pto::BindTileOp>())
+    return materializeScalarAccessPtr(bind.getSource(), rewriter, loc);
+
+  if (auto pointerCast = source.getDefiningOp<pto::PointerCastOp>()) {
+    if (pointerCast.getAddrs().empty())
+      return {};
+    Value addr = pointerCast.getAddrs().front();
+    if (isa<pto::PtrType>(addr.getType()))
+      return addr;
+    return materializeScalarAccessPtr(addr, rewriter, loc);
+  }
+
+  // Restrict normalization to memref views that already sit on top of a ptr-like
+  // boundary bridge. Materializing fresh memref -> ptr casts here would leave
+  // illegal pto.castptr(memref) behind in this pass.
+  return {};
 }
 
 struct ConvertTileBufAddrToPtrPattern
@@ -320,6 +386,47 @@ struct ConvertVstsSubviewOperandPattern : public OpConversionPattern<pto::VstsOp
   }
 };
 
+struct ConvertLoadScalarOperandToPtrPattern
+    : public OpConversionPattern<pto::LoadScalarOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(pto::LoadScalarOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value ptr = materializeScalarAccessPtr(adaptor.getPtr(), rewriter, op.getLoc());
+    if (!ptr)
+      return rewriter.notifyMatchFailure(op,
+                                         "failed to materialize load_scalar ptr");
+    if (!isa<pto::PtrType>(ptr.getType()))
+      return rewriter.notifyMatchFailure(op, "expected ptr-form load_scalar input");
+
+    rewriter.replaceOpWithNewOp<pto::LoadScalarOp>(op, op.getValue().getType(),
+                                                   ptr, adaptor.getOffset());
+    return success();
+  }
+};
+
+struct ConvertStoreScalarOperandToPtrPattern
+    : public OpConversionPattern<pto::StoreScalarOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(pto::StoreScalarOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value ptr = materializeScalarAccessPtr(adaptor.getPtr(), rewriter, op.getLoc());
+    if (!ptr)
+      return rewriter.notifyMatchFailure(op,
+                                         "failed to materialize store_scalar ptr");
+    if (!isa<pto::PtrType>(ptr.getType()))
+      return rewriter.notifyMatchFailure(op, "expected ptr-form store_scalar input");
+
+    rewriter.replaceOpWithNewOp<pto::StoreScalarOp>(op, ptr,
+                                                    adaptor.getOffset(),
+                                                    adaptor.getValue());
+    return success();
+  }
+};
+
 struct ConvertPtrNormalizeUnrealizedCastOp final
     : public OpConversionPattern<UnrealizedConversionCastOp> {
   using OpConversionPattern::OpConversionPattern;
@@ -403,6 +510,10 @@ struct VPTOPtrNormalizePass
     target.addDynamicallyLegalOp<pto::VstsOp>([](pto::VstsOp op) {
       return isa<pto::PtrType>(op.getDestination().getType());
     });
+    target.addDynamicallyLegalOp<pto::LoadScalarOp>(
+        [](pto::LoadScalarOp op) { return isa<pto::PtrType>(op.getPtr().getType()); });
+    target.addDynamicallyLegalOp<pto::StoreScalarOp>(
+        [](pto::StoreScalarOp op) { return isa<pto::PtrType>(op.getPtr().getType()); });
     target.addDynamicallyLegalOp<memref::SubViewOp>(
         [](memref::SubViewOp op) { return !needsSubviewPtrConversion(op); });
 
@@ -416,6 +527,8 @@ struct VPTOPtrNormalizePass
                  ConvertBindTileToPtrPattern,
                  ConvertSubviewToAddPtrPattern, ConvertVldsSubviewOperandPattern,
                  ConvertVstsSubviewOperandPattern,
+                 ConvertLoadScalarOperandToPtrPattern,
+                 ConvertStoreScalarOperandToPtrPattern,
                  ConvertPtrNormalizeUnrealizedCastOp>(
         typeConverter, context);
 

--- a/test/basic/issue_247_load_scalar_ptr_normalize.pto
+++ b/test/basic/issue_247_load_scalar_ptr_normalize.pto
@@ -1,0 +1,26 @@
+// Copyright (c) 2026 Huawei Technologies Co., Ltd.
+// This program is free software, you can redistribute it and/or modify it under the terms and conditions of
+// CANN Open Software License Agreement Version 2.0 (the "License").
+// Please refer to the License for details. You may not use this file except in compliance with the License.
+// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
+// See LICENSE in the root of the software repository for the full text of the License.
+
+// RUN: ptoas --pto-arch=a5 --pto-backend=vpto --emit-vpto --mlir-print-ir-after=vpto-ptr-normalize %s -o /dev/null 2>&1 | FileCheck %s --check-prefix=NORMALIZE
+// RUN: ptoas --pto-arch=a5 --pto-backend=vpto --vpto-emit-hivm-llvm %s -o /dev/null
+
+module attributes {pto.target_arch = "a5"} {
+  func.func @load_scalar_ui32(%ptr: !pto.ptr<ui32>) {
+    %c0 = arith.constant 0 : index
+    %v = pto.load_scalar %ptr[%c0] : !pto.ptr<ui32> -> ui32
+    pto.store_scalar %v, %ptr[%c0] : !pto.ptr<ui32>, ui32
+    return
+  }
+}
+
+// NORMALIZE: // -----// IR Dump After
+// NORMALIZE-SAME: (vpto-ptr-normalize)
+// NORMALIZE-LABEL: func.func @load_scalar_ui32(%arg0: !pto.ptr<ui32, gm>) {
+// NORMALIZE: %[[C0:.*]] = arith.constant 0 : index
+// NORMALIZE: %[[VAL:.*]] = pto.load_scalar %arg0[%[[C0]]] : !pto.ptr<ui32, gm> -> ui32
+// NORMALIZE: pto.store_scalar %[[VAL]], %arg0[%[[C0]]] : !pto.ptr<ui32, gm>, ui32

